### PR TITLE
refactor: move playerDetectTime from Monster to TracePlayer

### PR DESCRIPTION
**Pull request by MinecraftDawn**
I also changed the detectPlayerTime calculation logic

### CODE REVIEW

**Comment by MinecraftDawn** [comment]
I want to refact the time calculation as a separate class, but not in this PR.

File: Assets/Scripts/Strategy/TracePlayer.cs [hunk 1]
```diff
@@ -9,24 +9,42 @@ namespace Strategy
 public class TracePlayer : KeepMove
 {
     public float speedUpWhenTracing = 1.5f;
+    public float playerDetectTime = 3f;
     public float tracePlayerTurnAroundDelay = 0.8f;
-    
+    public Detector playerDetector;
```
Comment (DerrickPikachu) [inline Assets/Scripts/Strategy/TracePlayer.cs:14]: In my opinion, I would like to keep Detector and AI independent.
If AI only have coupling with monster that would be the best.
This version of AI make the detector have dependency with it which means AI have coupling with monster and Detector now.

**Review by DerrickPikachu — COMMENTED** [review]
I have some comment, please check.
You can call me if you have some concern.

File: Assets/Scripts/Strategy/TracePlayer.cs [hunk 2]
```diff
@@ -27,13 +26,15 @@ public override void Decide(Monster monster)
...
+        GameObject? player = monster.GetDetectedPlayer();
+        if (player is not null) {
+            _player = player;
             UpdateLastDetectPlayerTime();
         }
-
-        if (IsForgetPlayer()) { _player = null; }
+        else if(IsForgetPlayer()) {
+            _player = null;
+        }
```
Comment (DerrickPikachu) [inline Assets/Scripts/Strategy/TracePlayer.cs:37]: Why not "else if" just follow "}"?

File: Assets/Scripts/Monster.cs [hunk 3]
```diff
@@ -28,6 +29,10 @@ protected override void PreparationBeforeFixedUpdate()
         CollectState();
         if (_strategy != null) { _strategy.Decide(this); }
     }
+    
+    public GameObject GetDetectedPlayer() {
+        return detector.GetDetectedObject();
+    }
```
Comment (DerrickPikachu) [inline Assets/Scripts/Monster.cs:35]: This public method should check whether this monster have detector or not as I had mentioned in the meeting.

**Review by DerrickPikachu — COMMENTED** [review]
two more comment.

**Pull request merged**